File: sdk_swift/Tests/AllyourbaseTests/Support/MockSSETransport.swift
import Foundation
@testable import Allyourbase

final class MockSSEConnection: SSEConnection {
    private let lock = NSLock()
    private var continuation: AsyncThrowingStream<UInt8, Error>.Continuation?
    private(set) var cancelled = false

    func byteStream() -> AsyncThrowingStream<UInt8, Error> {
        AsyncThrowingStream { continuation in
            lock.lock()
            self.continuation = continuation
            let isCancelled = cancelled
            lock.unlock()
            if isCancelled {
                continuation.finish()
            }
        }
    }

    func cancel() {
        lock.lock()
        cancelled = true
        let continuation = self.continuation
        lock.unlock()
        continuation?.finish()
    }

    func send(text: String) {
        let bytes = Array(text.utf8)
        lock.lock()
        let continuation = self.continuation
        lock.unlock()
        for byte in bytes {
            continuation?.yield(byte)
        }
    }

    func finish() {
        lock.lock()
        let continuation = self.continuation
        lock.unlock()
        continuation?.finish()
    }
}

final class MockSSETransport: SSETransport {
    enum Behavior {
        case connect(MockSSEConnection)
        case fail(Error)
    }

    private(set) var requests: [HTTPRequest] = []
    private var queue: [Behavior] = []

    func enqueue(connection: MockSSEConnection) {
        queue.append(.connect(connection))
    }

    func enqueue(error: Error) {
        queue.append(.fail(error))
    }

    func connect(_ request: HTTPRequest) async throws -> any SSEConnection {
        requests.append(request)
        guard !queue.isEmpty else {
            throw MockTransportError.missingResponse
        }
        let next = queue.removeFirst()
        switch next {
        case let .connect(connection):
            return connection
        case let .fail(error):
            throw error
        }
    }
}

File: sdk_kotlin/src/test/kotlin/dev/allyourbase/MockSseTransport.kt
package dev.allyourbase

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.awaitCancellation
import kotlinx.coroutines.flow.flow
import java.util.concurrent.atomic.AtomicBoolean

sealed class MockSseBehavior {
    data class Connect(val lines: List<String>, val keepOpen: Boolean) : MockSseBehavior()
    data class Fail(val error: Throwable) : MockSseBehavior()
}

class MockSseTransport : SseTransport {
    val requests: MutableList<HttpRequest> = mutableListOf()
    val connections: MutableList<MockSseConnection> = mutableListOf()
    private val queue: ArrayDeque<MockSseBehavior> = ArrayDeque()

    fun enqueue(lines: List<String>, keepOpen: Boolean = false) {
        queue.addLast(MockSseBehavior.Connect(lines, keepOpen))
    }

    fun enqueue(error: Throwable) {
        queue.addLast(MockSseBehavior.Fail(error))
    }

    override suspend fun connect(request: HttpRequest): SseConnection {
        requests.add(request)
        return when (val next = queue.removeFirstOrNull() ?: throw IllegalStateException("MockSseTransport queue is empty")) {
            is MockSseBehavior.Connect -> MockSseConnection(next.lines, next.keepOpen).also { connections.add(it) }
            is MockSseBehavior.Fail -> throw next.error
        }
    }
}

class MockSseConnection(
    private val sourceLines: List<String>,
    private val keepOpen: Boolean = false,
) : SseConnection {
    private val cancelled = AtomicBoolean(false)
    val isCancelled: Boolean
        get() = cancelled.get()

    override fun lines(): Flow<String> = flow {
        for (line in sourceLines) {
            if (cancelled.get()) {
                break
            }
            emit(line)
        }
        if (keepOpen && !cancelled.get()) {
            try {
                awaitCancellation()
            } catch (_: Throwable) {
                // noop
            }
        }
    }

    override fun cancel() {
        cancelled.set(true)
    }
}
